actors/builtin/miner/miner_actor.go

```diff
@@ -263,6 +271,8 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
 		provenSectors, err := abi.BitFieldUnion(partitionsSectors...)
 		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to union %d partitions of sectors", len(partitionsSectors))
 
+		// Load sector infos, substituting a known-good sector for known-faulty sectors.
+		// XXX: ErrIllegalState isn't quite correct. This can fail if the user submits a proof for a partition with only faulty sectors.
```

Comment:
?

Reply:
Yeah, this is an awkwardness we have currently that errors propagated up from below can't indicate what the appropriate exit code is (but, OTOH, sometimes the bottom level doesn't know and the decision should be made here).

A solution probably involves a custom error type that can carry an exit code and message, and using that exit code if provided. Or some other error-inspection mechanism.

This isn't the only place, and I'd welcome generally your thoughts on how to do this better. P2 though.

Reply:
We can introduce a custom error interface:

```go
type ExitError interface {
    ErrorCode() ExitCode
}
```

Then, in `builtin.RequireNoError`, we'd use the error code argument as the _default_ error code, first trying to extract an error code from the error itself. Of course, that could cause it's own problems and bugs.

Ideally, we'd be able to check arguments up-front and avoid this mess, but that's not always possible.